### PR TITLE
adding check for matching signing algorithm of token

### PR DESCRIPTION
**Pull request by gitu**
solves #228

### CODE REVIEW

**Comment by gitu** [comment]
@simongottschlag can you have a look please

File: internal/oidc/keyhandler.go [hunk 1]
```diff
@@ -133,26 +133,39 @@ func (h *keyHandler) getKeySet() jwk.Set {
 	return h.keySet
 }
 
-func (h *keyHandler) getKeyFromID(ctx context.Context, keyID string) (jwk.Key, error) {
+func (h *keyHandler) getKeyFromID(ctx context.Context, keyID string, algorithm jwa.SignatureAlgorithm) (jwk.Key, error) {
 	keySet := h.getKeySet()
 
+	key, err := findKey(ctx, keyID, algorithm, keySet)
+	if err == nil {
+		return key, err
```
Comment (simongottschlag) [inline internal/oidc/keyhandler.go:141]: error is nil, but returning error (linter will warn).

```suggestion
		return key, nil
```

File: internal/oidc/keyhandler.go [hunk 2]
```diff
@@ -3,6 +3,7 @@ package oidc
 import (
 	"context"
 	"fmt"
+	"github.com/lestrrat-go/jwx/jwa"
 	"net/http"
 	"sync"
 	"time"
```
Comment (simongottschlag) [inline internal/oidc/keyhandler.go:9]: Place the non-std lib import as follows:

```suggestion
	"context"
	"fmt"
	"net/http"
	"sync"
	"time"
	
	"github.com/lestrrat-go/jwx/jwa"
```

File: internal/oidc/keyhandler_test.go [hunk 3]
```diff
@@ -77,7 +83,7 @@ func TestNewKeyHandler(t *testing.T) {
 	keyID3, err := getKeyIDFromTokenString(token3.AccessToken)
 	require.NoError(t, err)
 	op.Close(t)
-	_, err = keyHandler.getKeyFromID(ctx, keyID3)
+	_, err = keyHandler.getKeyFromID(ctx, keyID3, "")
 	require.Error(t, err)
 }
```
Comment (simongottschlag) [inline internal/oidc/keyhandler_test.go:88]: I would like to add a new test that has two keys with the same `kid` but different `alg` and make sure this works and continues to work in the future.

**Review by simongottschlag — CHANGES_REQUESTED** [review]
Hi @gitu!

Thank you so much for your contribution. Amzing work. :1st_place_medal: 

I've added a few comments to the PR. Please tell me if you don't agree with something and we'll take the conversation.

One important thing before merge will be to have a test that verifies your issue in `keyhandler_test.go`.

I may be a bit slow this and next week, but I'll do my best to continue the review as soon as possible when you're ready.

File: internal/oidc/oidc.go [hunk 4]
```diff
@@ -280,6 +285,20 @@ func getKeyIDFromTokenString(tokenString string) (string, error) {
 	return keyID, nil
 }
 
+func getTokenAlgorithmFromTokenString(tokenString string) (jwa.SignatureAlgorithm, error) {
+	headers, err := getHeadersFromTokenString(tokenString)
```
[the remark this answers is not in the document]
Comment (gitu) [reply internal/oidc/oidc.go]: I agree, did so.

**Comment by gitu** [comment]
@simongottschlag thanks for your input, feel free to have a look again

**Comment by simongottschlag** [comment]
@gitu I will be merging this into a separate branch and add a few fixup commits (for linting etc). Will ping you in the other PR and do my best not to screw up the merge to make sure your commits are pointing to you. :^)

**Pull request merged**